node/src/api/registration.rs

```diff
@@ -1,181 +1,118 @@
-            data: Some(sender_json),
-            type_: Some(String::from("sender")),
-        };
-        let post_request = RegistrationapiResourcePostRequest::Variant2(sender_post_request);
-
-        client.post(url.clone()).json(&post_request).send().await?;
-
-        Ok(())
+        resource: &dyn resource::Registerable,
+    ) -> Result<reqwest::Response, Box<dyn std::error::Error>> {
+        let delete_url = url
+            .clone()
+            .join(format!("resource/{}", resource.registry_path()).as_str())
+            .unwrap();
+
+        Ok(client.delete(delete_url).send().await?.error_for_status()?)
     }
 
-    async fn register_receiver(
+    pub async fn register_resource(
         client: &reqwest::Client,
         url: &reqwest::Url,
-        receiver: &resource::Receiver,
-    ) -> Result<(), Box<dyn std::error::Error>> {
-        use nmos_model::version::is_04::V1_0;
-        use nmos_schema::is_04::v1_0_x::{
-            RegistrationapiResourcePostRequest, RegistrationapiResourcePostRequestHealthVariant3,
-        };
-
-        let resource::ReceiverJson::V1_0(receiver_json) = receiver.to_json(&V1_0);
-        let receiver_post_request = RegistrationapiResourcePostRequestHealthVariant3 {
-            data: Some(receiver_json),
-            type_: Some(String::from("receiver")),
-        };
-        let post_request = RegistrationapiResourcePostRequest::Variant3(receiver_post_request);
-
-        client.post(url.clone()).json(&post_request).send().await?;
+        resource: &dyn resource::Registerable,
+        api_version: &APIVersion,
+    ) -> Result<reqwest::Response, Box<dyn std::error::Error>> {
+        let res = Self::post_resource(client, url, resource, api_version).await?;
+
+        if res.status() == StatusCode::OK {
+            warn!(
+                "Resource already present in API deleting: {}",
+                resource.registry_path()
+            );
+
+            let res = Self::delete_resource(client, url, resource).await?;
+
+            if res.status() == StatusCode::OK {
+                return Err(Box::new(std::io::Error::new(
+                    std::io::ErrorKind::Other,
+                    "Failed to register resource after deleting and re-registering",
+                )));
+            } else {
+                return Ok(res);
+            }
+        }
 
-        Ok(())
+        Ok(res)
     }
 
     pub async fn register_resources(
         client: &reqwest::Client,
         model: Arc<Model>,
-        registry: &NmosMdnsRegistry,
+        registry: Arc<Mutex<Option<NmosMdnsRegistry>>>,
+        api_version: &APIVersion,
     ) -> Result<(), Box<dyn std::error::Error>> {
-        let base = &registry.url.join("v1.0/").unwrap();
+        let registry = registry.lock().await.clone().unwrap();
 
-        info!("Attempting to register with {}", base);
+        let base = &registry
+            .url
+            .join(format!("{}/", api_version.to_string()).as_str())
```

Comment:
```suggestion
            .join(format!("{}/", api_version).as_str())
```
How about implementing Display based on to_string for API version?